File: code_snippets/kotlin/src/main/kotlin/usecases/eventprocessing/UserFeed.kt
package usecases.eventprocessing

import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.VirtualObject
import dev.restate.sdk.endpoint.Endpoint
import dev.restate.sdk.kotlin.*
import dev.restate.sdk.lambda.BaseRestateLambdaHandler
import kotlin.time.Duration.Companion.seconds
import kotlinx.serialization.Serializable
import usecases.microservices.vo.SubscriptionObject

// <start_here>
@VirtualObject
class UserFeed {

  @Serializable data class SocialMediaPost(val content: String, val metadata: String)

  // <mark_1>
  @Handler
  suspend fun processPost(ctx: ObjectContext, post: SocialMediaPost) {
    // </mark_1>
    // <mark_5>
    val userId = ctx.key()
    // </mark_5>

    // <mark_3>
    val postId = ctx.runBlock { createPost(userId, post) }
    // </mark_3>

    // <mark_4>
    // <mark_3>
    while (ctx.runBlock { getPostStatus(postId) } == "PENDING") {
      // </mark_3>
      // <mark_2>
      ctx.sleep(5.seconds)
      // </mark_2>
    }
    // </mark_4>

    // <mark_3>
    ctx.runBlock { updateUserFeed(userId, postId) }
    // </mark_3>
  }
}

// <mark_6>
class MyLambdaHandler : BaseRestateLambdaHandler() {
  override fun register(builder: Endpoint.Builder) {
    builder.bind(SubscriptionObject())
  }
}
// </mark_6>
// <end_here>

fun createPost(userId: String, post: UserFeed.SocialMediaPost): String {
  // Implementation here
  return "postId"
}

fun getPostStatus(postId: String): String {
  // Implementation here
  return "status"
}

fun updateUserFeed(userId: String, postId: String) {
  // Implementation here
}

File: code_snippets/python/src/use_cases/event_processing/user_feed.py
from datetime import timedelta

import restate
from restate import VirtualObject, ObjectContext

from src.use_cases.event_processing.utils import (
    create_post,
    get_post_status,
    update_user_feed,
    SocialMediaPost,
    Status,
)

# <start_here>
user_feed = VirtualObject("UserFeed")


# <mark_1>
@user_feed.handler()
async def process_post(ctx: ObjectContext, post: SocialMediaPost):
    # </mark_1>
    # <mark_5>
    user_id = ctx.key()
    # </mark_5>

    # <mark_3>
    post_id = await ctx.run("profile update", lambda: create_post(user_id, post))
    # </mark_3>

    # <mark_4>
    # <mark_3>
    current_status = await ctx.run("post status", lambda: get_post_status(post_id))
    while current_status == Status.PENDING:
        # </mark_3>
        # <mark_2>
        await ctx.sleep(timedelta(seconds=5))
        # </mark_2>
    # </mark_4>

    # <mark_3>
    await ctx.run("update feed", lambda: update_user_feed(user_id, post_id))
    # </mark_3>


# <mark_6>
aws_lambda_handler = restate.app([user_feed])
# </mark_6>
# <end_here>
